lms/djangoapps/course_api/blocks/transformers/milestones.py

```diff
@@ -0,0 +1,84 @@
+"""
+Milestones Transformer
+"""
+
+from django.conf import settings
+
+from edx_proctoring.api import get_attempt_status_summary
+from edx_proctoring.models import ProctoredExamStudentAttemptStatus
+from openedx.core.lib.block_structure.transformer import BlockStructureTransformer, FilteringTransformerMixin
+from util import milestones_helpers
+
+
+class MilestonesTransformer(FilteringTransformerMixin, BlockStructureTransformer):
+    """
+    Exclude proctored exams unless the user is not a verified student or has
+    declined taking the exam.
+    """
+    VERSION = 1
+    BLOCK_HAS_PROCTORED_EXAM = 'has_proctored_exam'
+
+    @classmethod
+    def name(cls):
+        return "milestones"
+
+    @classmethod
+    def collect(cls, block_structure):
+        """
+        Computes any information for each XBlock that's necessary to execute
+        this transformer's transform method.
+
+        Arguments:
+            block_structure (BlockStructureCollectedData)
+        """
+        block_structure.request_xblock_fields('is_proctored_enabled')
+        block_structure.request_xblock_fields('is_practice_exam')
+
+    def transform_block_filters(self, usage_info, block_structure):
+        def user_gated_from_block(block_key):
+            """
+            Checks whether the user is gated from accessing this block, first via exam proctoring, then via a general
+            milestones check.
+            """
+            return settings.FEATURES.get('ENABLE_SPECIAL_EXAMS', False) and not usage_info.has_staff_access and (
+                self.is_proctored_exam(block_key, usage_info, block_structure) or
+                self.has_pending_milestones_for_user(block_key, usage_info)
+            )
+
+        return [block_structure.create_removal_filter(user_gated_from_block)]
+
+    @staticmethod
+    def is_proctored_exam(block_key, usage_info, block_structure):
+        """
+        Test whether the block is a special exam for the user in
```

Comment:
Update docstring to say proctored exam and practice-proctored exam, since it doesn't include all special exams.